0017-sampler-interface.md

```diff
@@ -0,0 +1,248 @@
+
+However, a `uint8` array alone is insufficient because, at the very least, it is incapabable of specifying whether it describes a bit-array register of size 14 or 16, for example, both of which require two bytes.
+For this reason, and to handle convenience methods, we introduce the `BitArray`.
+A bare bones sketch is below.
+Methods like `__or__` or static constructors from other formats are in-scope of the vision, but discluded for brevity.
+
+```python
+class BitArray(ShapedMixin):
+    """Stores bit-array outcomes."""
+
+    def __init__(self, array: NDArray, num_bits: int):
+        """
+        Args:
+            array: The data, where the last axis is over packed bits whose byte order is 
+                big-endian, the second last axis is over shots, and the preceding axes 
+                correspond to the shape of the experiment.
+            num_bits: How many bit are in each outcome; the size of the bit-array register.
+        """
+        self._array = np.array(array, copy=False, dtype=np.uint8)
+        if self._array.ndim < 2:
+            raise ValueError("The data arary needs to be at least 2-D")
+        self._num_bits = num_bits
+        if self._array.shape[-1] * 8 < self._num_Bits:
+            raise ValueError("The last axis of the data array is not big enough.")
+        # second last dimension is shots, last dimension is packed bits
+        self._shape = self._array.shape[:-2]
+
+    def __repr__(self):
+        desc = f"<num_samples={self.num_shots}, num_bits={self.num_bits}, shape={self.shape}>"
+        return f"BitArray({desc})"
+
+    @property
+    def array(self) -> NDArray[np.uint8]:
+        """The raw array of data."""
+        return self._array
+
+    @property
+    def num_bits(self) -> int:
+        """The number bits in each outcome."""
+        return self._num_bits
+
+    @property
+    def num_samples(self) -> int:
+        """The number of samples."""
+        return self._array.shape[-2]
+
+    def get_counts(self, loc: Tuple[int,...] | None = None, marginal_idxs: Tuple[int,...] | None = None) -> Mapping[str, int]:
+            def get_counts(self, idx: Tuple[int, ...] | None = None) -> Counts:
+        """Join all bit-array output registers together into a new counts structure.
+        
+        Args:
+            idx: Which task coordinate to fetch.
+
+        Raises:
+            ValueError: If no ``idx`` is given and this task result is more than 0-D.
+        """
+```
+
+## Migration Path <a name="migration-path"></a>
+
+An analagous path as [the paired Estimator RFC](https://github.com/Qiskit/RFCs/pull/51). Convenience methods can be placed on result containers to transform into familiar return types.
+
+## Questions <a name="questions"></a>
+
+ * Should `num_samples` (i.e. `shots`) be an argument of the `Task`? Having ditched outcome weights, this seems natural.
```

Comment:
We decided to keep `shots` as an execution option, and not part of `EstimatorTask`. We decided that some sort of precision argument could be appropriate inside of the task in a future iteration.

Reply:
I guess I am not against allowing for per-task `shots` arguments as long as we also have a mechanism for specifying uniform `shots` over all the tasks, which I think is the most common case.

I will note that non-uniform shots certainly has downstream implications for how we execute, because today we make the shots the "outer" loop after sequentially executing each circuit in calls to `backend.run(...)`. There is nothing about the legacy interface which implies that ordering, it is just what we do today.

Reply:
[fed002d](https://github.com/ihincks/qiskit-rfcs/pull/9/commits/fed002de00f7cfe06576c22daa4100911773e68e)